### DISCUSSION TITLE
Why doesn't airflow 2 run queued jobs in order?

### DISCUSSION THREAD

**Discussion by HashSalmon** [apache/airflow#20177] [Q&A]
Hey Guys, I have a question regarding the executing sequence of queued jobs.
Some Envs:
 - airflow 2.1.2
 - executor: CeleryExecutor
 - concurrency=2,
 - max_active_runs=2,
 
Let's say I manually triggered 10 jobs from POSTMAN by calling the airflow API. I have concurrency=2 and max_active_runs=2 in the DAG decorator. On the UI, I am expecting the airflow executes the jobs from left to right with 2 running jobs at the same time. However, it executes jobs in a random order. Please see the attached graph for a reference.

I am expecting airflow to execute job1 and job2 at the same time initially, then job3, job4.. But it seems like executes the jobs not in order. Please let me know if there is a way to fix it. Thanks!

![Screen Shot 2021-12-09 at 11 43 41 AM](https://user-images.githubusercontent.com/2868104/145465167-3857cd82-be74-46b9-9a1f-c33c141837cc.png)

**Comment by potiuk** [comment]
I wonder why did you expect that to happen? What was your expectation based on :) ?

Conceptually in Airflow each task is independent from each other and nowhere it is forced that the tasks should execute in specific order (unless there is a dependency).  And we are not forcing that (which allows for a number of optimisations - especially when you have mutliple schedulers, trying to execute tasks in time order "in general" would impact the scalability of the solution. So there is no implicit "execute this job before that job". Airflow simply schedules stuff when it sees they can be scheduled and indeed picks pretty much random tasks from the list of tasks that are eligible for execution.

However I think what you are looking for can be achieved and Airflow has tools for that. Make sure that last task in your DAG depenc on your first task and set  `wait_for_downstream=True` for the fist task. This means that your task will wait for all "downstram" tasks of previous execution of it. 

See more details here:
https://stackoverflow.com/questions/41009228/is-it-possible-for-airflow-scheduler-to-first-finish-the-previous-days-cycle-be

**Reply by HashSalmon** [reply]
Hey, thanks for the help. I thought the jobs were put into the queue one by one. That was why I expected them to be executed in order.
Let me try the suggestion and will get back to you for the result. Thanks!

**Reply by potiuk** [reply]
There is no "queue" in Airflow. That would be impossible (or very difficult) to scale. There is just a pool of tasks that are "eligible to run" and the only way to make some task not run is to make them "not eligible to run" :). Which the solution I described is about :)
